src/main/scala/piconot/external/parser/Parser.scala

```diff
@@ -0,0 +1,130 @@
+package piconot.external.parser
+
+import scala.language.postfixOps
+import scala.util.parsing.combinator.{PackratParsers, RegexParsers}
+import picolib.semantics._
+
+/*
+ * Syntax for a valid Picobot rule:
+ *
+ *         stmts ::= "while" "(" expr ")" body | "move" dir
+ *          expr ::= "free" dir | "blocked" dir
+ *           dir ::= "N" | "S" | "E" | "W"
+ *          body ::= "{" body | stmts body | "}"
+ *
+ */
+
+object PiconotParser extends RegexParsers{// with PackratParsers{
+  // carried over from internal
+  implicit def intToState(i: Int): State = State(i.toString)
+  implicit def stateToInt(s: State): Int = s.name.toInt
+  def anySurr: Surroundings = Surroundings(Anything, Anything, Anything, Anything)
+
+  // helper function - returns the opposite relative description of surrounding
+  def opp(r: RelativeDescription): RelativeDescription = r match {
+    case Open => Blocked
+    case Blocked => Open
+    case Anything => Anything
+  }
+
+  // returns the "opposite surroundings"
+  def opposite(surr: Surroundings): Surroundings = {
+    Surroundings(opp(surr.north), opp(surr.east), opp(surr.west), opp(surr.south))
+  }
+
+  // converts to surroundings with one open direction
+  def free(dir: MoveDirection): Surroundings = dir match {
+    case North => Surroundings(Open, Anything, Anything, Anything)
+    case East => Surroundings(Anything, Open, Anything, Anything)
+    case West => Surroundings(Anything, Anything, Open, Anything)
+    case South => Surroundings(Anything, Anything, Anything, Open)
+  }
+
+  // for parsing comments
+  override protected val whiteSpace =
+  """(\s|#.*)+""".r
+
+  // parsing interface
+  def apply(s: String): ParseResult[List[List[Rule]]] = {
+    parseAll(program, s)
+  }
+
+  def program: Parser[List[List[Rule]]] = stmts *
+
+  def stmts: Parser[List[Rule]] =
+    (("while" ~ "(" ~ expr ~ ")" ~ body ^^ { case "while" ~ "(" ~ condition ~ ")" ~ b =>
+          fixState(List(Rule(0, condition, StayHere, 1)) ++ List(Rule(0, opposite(condition), StayHere, 1)) ++ b)  })
+      | ("move" ~ dir ^^ { case "move" ~ d => List(Rule(0, anySurr, d, 1)) })
+      | failure("expected a 'while' or 'move' statement")
+    )
+
+  def expr: Parser[Surroundings] =
+    (("free" ~ dir ^^ { case "free" ~ d => free(d) })
+      | ("blocked" ~ dir ^^ { case "blocked" ~ d => opposite(free(d)) })
+      | ("any" ^^^ anySurr)
+      | failure("expected 'free', 'blocked', or 'any' expression")
```

Comment:
Maybe you could also have an error message for when a user enters a MoveDirection other than N, E, W, or S? We didn't get to that, but it was something that we wanted to do.